### PR TITLE
[Slider] Allowing Full Haptics for Discrete Sliders

### PR DESCRIPTION
**Pull request by afweiss**
This adds the ability to add full haptics for discrete sliders. Full haptics are opt-in. When full haptics are enabled, a light vibration occurs whenever the slider changes values.

### CODE REVIEW

File: components/Slider/src/MDCSlider.h [hunk 1]
```diff
@@ -335,13 +335,21 @@ IB_DESIGNABLE
...
  */
 @property(nonatomic, assign) BOOL hapticsEnabled;
 
+/** When @c YES, haptics for any value change are enabled for discrete sliders. The haptics casue
+ a light impact reaction when the slider value changes for discrete sliders. Can only be set to yes
+ for discrete sliders.
+
+ Defaults to @c NO
+ */
+@property(nonatomic, assign) BOOL fullHapticsEnabled;
```
Comment (yarneo) [inline components/Slider/src/MDCSlider.h]: I feel that the naming "fullHapticsEnabled" may be ambiguous to some users. I wonder if we should be more explicit about what is being enabled when this property is set to YES. I would try to see how Apple names similar properties that create similar behaviors.

One suggestion would be to name the property `shouldEnableHapticsForAllDiscreteValues`. In Objective-C it's recommended for properties to be "self-documenting".

File: components/Slider/src/MDCSlider.m [hunk 2]
```diff
@@ -544,8 +553,10 @@ - (void)thumbTrackValueChanged:(__unused MDCThumbTrack *)thumbTrack {
   [self sendActionsForControlEvents:UIControlEventValueChanged];
   UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, self.accessibilityValue);
   if (@available(iOS 10.0, *)) {
-    if (self.hapticsEnabled && (_thumbTrack.value == _thumbTrack.minimumValue ||
-                                _thumbTrack.value == _thumbTrack.maximumValue)) {
+    if (self.fullHapticsEnabled) {
```
Comment (yarneo) [inline components/Slider/src/MDCSlider.m]: what happens if hapticsEnabled is set to NO, do we still want haptics to occur if fullhapticsenabled is set to YES?

File: components/Slider/src/MDCSlider.m [hunk 3]
```diff
@@ -332,6 +333,14 @@ - (void)setHapticsEnabled:(BOOL)hapticsEnabled {
   }
 }
 
+- (void)setFullHapticsEnabled:(BOOL)fullHapticsEnabled {
+  if (@available(iOS 10.0, *)) {
+    if (_thumbTrack.numDiscreteValues >= 2) {
```
Comment (yarneo) [inline components/Slider/src/MDCSlider.m:338]: what happens if the number of discrete values is 1?

File: components/Slider/tests/unit/SliderTests.m [hunk 4]
```diff
@@ -1166,6 +1195,40 @@ - (void)testNotEnabledHapticFeedback {
   }
 }
 
+- (void)testEnabledFullHapticFeedback {
+  // Given
+  self.slider.minimumValue = 0;
+  self.slider.maximumValue = 5;
+  self.slider.hapticsEnabled = NO;
```
Comment (yarneo) [inline components/Slider/tests/unit/SliderTests.m:1202]: i would expect if hapticsEnabled is NO that no haptics would occur

File: components/Slider/tests/unit/SliderTests.m [hunk 5]
```diff
@@ -1166,6 +1195,40 @@ - (void)testNotEnabledHapticFeedback {
...
+
+      // Then
+      XCTAssertTrue(_mockFeedbackGenerator.impactHasOccurred);
+
+      _mockFeedbackGenerator.impactHasOccurred = NO;
+    }
+  }
+}
+
+- (void)testDefaultHapticsEnabledValues {
```
Comment (yarneo) [inline components/Slider/tests/unit/SliderTests.m]: Looks like this is not testing the new property added, is this related more to the previous PR?

**Review by yarneo — COMMENTED** [review]
Looks good overall! I left a few comments, curious to hear your thoughts. P.S. Naming is hard :)

File: components/Slider/src/MDCSlider.m (hunk 3 above)
Comment (afweiss) [reply components/Slider/src/MDCSlider.m:338]: "if numDiscreteValues is set to one, then the thumb track will act as if numDiscreteValues is zero and will judge you silently." - in MDCThumbTrack.h
I think if the number of discrete values is 1 then everything is treated as a continuous slider because a slider would need at least two discrete values.

Comment (yarneo) [reply components/Slider/src/MDCSlider.m:338]: awesome, thanks!

File: components/Slider/src/MDCSlider.h [hunk 6]
```diff
@@ -335,13 +335,21 @@ IB_DESIGNABLE
...
+/** When @c YES, haptics for min and max are enabled. The haptics casue a light impact reaction when
+ the slider reaches the minimum or maximum value.
 
  Defaults to @c YES in iOS 10 or later, @c NO otherwise
  */
 @property(nonatomic, assign) BOOL hapticsEnabled;
 
+/** When @c YES, haptics for any value change are enabled for discrete sliders. The haptics casue
+ a light impact reaction when the slider value changes for discrete sliders. Can only be set to yes
+ for discrete sliders. Haptics will only occur if hapticsEnable is also set to @c YES.
```
Comment (yarneo) [inline components/Slider/src/MDCSlider.h]: small typo here :`hapticsEnabled`

File: components/Slider/src/MDCSlider.m [hunk 7]
```diff
@@ -544,8 +553,10 @@ - (void)thumbTrackValueChanged:(__unused MDCThumbTrack *)thumbTrack {
   [self sendActionsForControlEvents:UIControlEventValueChanged];
   UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, self.accessibilityValue);
   if (@available(iOS 10.0, *)) {
-    if (self.hapticsEnabled && (_thumbTrack.value == _thumbTrack.minimumValue ||
-                                _thumbTrack.value == _thumbTrack.maximumValue)) {
+    if (self.shouldEnableHapticsForAllDiscreteValues && self.hapticsEnabled) {
```
Comment (yarneo) [inline components/Slider/src/MDCSlider.m]:
```
if (self.hapticsEnabled) {
  if (self.shouldEnableHapticsForAllDiscreteValues || _thumbTrack.value == _thumbTrack.minimumValue || _thumbTrack.value == _thumbTrack.maximumValue) {
    [self.feedbackGenerator impactOccurred];
  }
}
```

**Review by yarneo — APPROVED** [review]
